presto-hudi/src/main/java/com/facebook/presto/hudi/HudiFileSkippingManager.java

```diff
@@ -0,0 +1,312 @@
+    private final HoodieTableMetaClient metaClient;
+    private final HoodieTableType tableType;
+    private final HoodieTableMetadata metadataTable;
+
+    protected transient volatile Map<String, List<FileSlice>> allInputFileSlices;
+
+    private transient volatile SyncableFileSystemView fileSystemView;
+
+    public HudiFileSkippingManager(
+            List<String> partitions,
+            String spillableDir,
+            HoodieEngineContext engineContext,
+            HoodieTableMetaClient metaClient,
+            HoodieTableQueryType queryType,
+            Option<String> specifiedQueryInstant)
+    {
+        this.queryType = queryType;
+        this.specifiedQueryInstant = specifiedQueryInstant;
+        this.metaClient = metaClient;
+        HoodieMetadataConfig metadataConfig = HoodieMetadataConfig.newBuilder().enable(true).build();
+        this.tableType = metaClient.getTableConfig().getTableType();
+        this.metadataTable = HoodieTableMetadata
+                .create(engineContext, metadataConfig, metaClient.getBasePathV2().toString(), spillableDir, true);
+        prepareAllInputFileSlices(partitions, engineContext, metadataConfig, spillableDir);
+    }
+
+    private void prepareAllInputFileSlices(List<String> partitions, HoodieEngineContext engineContext, HoodieMetadataConfig metadataConfig, String spillableDir)
+    {
+        long startTime = System.currentTimeMillis();
+        HoodieTimeline activeTimeline = metaClient.reloadActiveTimeline();
+        Option<HoodieInstant> latestInstant = activeTimeline.lastInstant();
+        // build system view.
+        fileSystemView = FileSystemViewManager
+                .createViewManager(engineContext, metadataConfig,
+                        FileSystemViewStorageConfig.newBuilder().withBaseStoreDir(spillableDir).build(),
+                        HoodieCommonConfig.newBuilder().build(),
+                        () -> metadataTable)
+                .getFileSystemView(metaClient);
+        Option<String> queryInstant = specifiedQueryInstant.or(() -> latestInstant.map(HoodieInstant::getTimestamp));
+        if (tableType.equals(HoodieTableType.MERGE_ON_READ) && queryType.equals(HoodieTableQueryType.SNAPSHOT)) {
+            allInputFileSlices = partitions.stream().collect(Collectors.toMap(Function.identity(),
+                    partitionPath ->
+                            queryInstant.map(instant ->
+                                    fileSystemView.getLatestMergedFileSlicesBeforeOrOn(partitionPath, queryInstant.get())
+                                            .collect(Collectors.toList()))
+                                    .orElse(Collections.emptyList())));
+        }
+        else {
+            allInputFileSlices = partitions.stream().collect(Collectors.toMap(Function.identity(), partition ->
+                    queryInstant.map(instant ->
+                            fileSystemView.getLatestFileSlicesBeforeOrOn(partition, instant, true))
```

Comment:
we should always use getLatestMergedFileSlicesBeforeOrOn for all query types. We identified a bug sometime back for read optimized table. 

Ref: https://github.com/apache/hudi/pull/6664/files

Reply:
fixed